packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/data/ProjectSearchService.java

```diff
@@ -0,0 +1,205 @@
+
+		static PermissionJoin CreateProjectJoin() {
+			final PermissionJoin join = new PermissionJoin();
+			join.setName("project");
+			join.setParent(null);
+			return join;
+		}
+
+		static PermissionJoin CreatePermissionJoin(final UUID projectId) {
+			final PermissionJoin join = new PermissionJoin();
+			join.setName("permission");
+			join.setParent(projectId);
+			return join;
+		}
+	}
+
+	@Data
+	public static class ProjectDocument {
+
+		private final String type = "project";
+		private String userId; // the user id of who owns the project
+		private Boolean publicAsset; // whether the project is public or not
+		private Timestamp createdOn;
+		private Timestamp updatedOn;
+
+		// join field properties
+		private PermissionJoin permissionJoin;
+	}
+
+	@Data
+	public static class ProjectPermissionDocument {
+
+		private final String type = "permission";
+		private String userId; // the user id who holds the permission
+
+		// join field properties
+		private PermissionJoin permissionJoin;
+	}
+
+	/**
+	 * Setup the index and alias for the asset this service manages and ensure it is
+	 * empty
+	 *
+	 * @throws IOException If there is an error setting up the index and alias
+	 */
+	public void setupIndexAndAliasAndEnsureEmpty() throws IOException {
+		log.info("Setting up index {} and alias {}", getIndex(), getAlias());
+		final String index = getIndex();
+		try {
+			final String currentIndex = getCurrentAssetIndex();
+			// if another index exists, delete it.
+			if (!currentIndex.equals(index)) {
+				elasticService.deleteIndex(currentIndex);
+			}
+		} catch (final Exception e) {}
+		elasticService.createOrEnsureIndexIsEmpty(index);
+		elasticService.createAlias(index, getAlias());
+	}
+
+	/**
+	 * Teardown the index and alias for the asset this service manages
+	 *
+	 * @throws IOException If there is an error tearing down the index and alias
+	 */
+	public void teardownIndexAndAlias() throws IOException {
+		log.info("Tearing down index {}", getIndex());
+		final String index = getCurrentAssetIndex();
+		elasticService.deleteIndex(index);
+	}
+
+	public void indexProject(final Project project) throws IOException {
+		final ProjectDocument doc = new ProjectDocument();
+		doc.setUserId(project.getUserId());
+		doc.setPublicAsset(project.getPublicAsset());
+		doc.setCreatedOn(new Timestamp(System.currentTimeMillis()));
+		doc.setUpdatedOn(new Timestamp(System.currentTimeMillis()));
+		doc.setPermissionJoin(PermissionJoin.CreateProjectJoin());
+
+		final String routing = project.getId().toString();
+
+		elasticService.indexWithRouting(getAlias(), project.getId().toString(), doc, routing);
+	}
+
+	public void updateProject(final Project project) throws IOException {
+		final ProjectDocument doc = new ProjectDocument();
+		doc.setUserId(project.getUserId());
+		doc.setPublicAsset(project.getPublicAsset());
+		doc.setUpdatedOn(new Timestamp(System.currentTimeMillis()));
+		doc.setPermissionJoin(PermissionJoin.CreateProjectJoin());
+
```

Comment:
Will this not erase the existing `PermissionJoin` thing?

Reply:
Or, I don't understand this data structure.  It appears that there is one PermissionJoin for a project and yet it changes to be either a specific user via `addProjectPermission` or reset to whatever `CreateProjectJoin()` creates...

Reply:
The two types of documents share the same index. So the "parent", which is the project, has a join field declaring it a project parent type. And the "child", which is the user read permission, has a join field declaring it a child doc. The `userId` field is overrided, in projects it is the owner of the project, and in a permission it is the userId of the person who has read permission.

Reply:
The structure of join fields is not very intuitive. They aren't separate docs in different indices, they are in the same index, and have the same mapping. The implicit `name` and `parent` property of the join field are used to define the relationship

Reply:
Well based on the created_on fiasco... I suspect this is where this code will fail